From a296ff757c97f2cbdb5e02af4d3e83cd9f7ba39b Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Wed, 15 Jun 2011 14:08:03 -0600
Subject: [PATCH] RFC: Make all headers from the PatchScreen available as popups.

Make the CommitMessage, the Preferences, the Patch Set Browser,
along with the FileList (already possible previously) available
as popups.  To make them available, simply click on the
Differences Menu for that particular item while it is already
displayed in the header.  At this point it disappears from the
header and appears as a popup which does not auto hide.  To
hide it, click on the menu entry again, and it will revert
back to the header.

This is a rough implementation to float the idea forward.  I
have not messed with any bindings, so they don't work right
yet, I welcome suggestions on whether these popups should get
bindings and what they should be.  I feel like ESC should be
used to close the popups, but I am not sure if that would
interfere with anything else?

Also, since there is no autohide, it might be nice to figure
out a way to close the window by clicking in it.  An X on the
right?

Another visible change is the height, I chose to reduce it to
1/3 max, it appeared to by set to 2/3rds previously. My
rational is that it used to be the only way to see the
filelist and 2/3rds made sense. But now with these available
in the header, full size, I feel like the popup should be
less intrusive.

Lastly, I noticed that the FileList uses the patchset as the
title, perhaps that should be pushed to the page title
(to replace the removed "Change SHA1:") instead in the future.
Should the popup header be blank (no title) to further save
space vertically?

Change-Id: I493701f0e250509ef253cf5fef72e40c8b65aba2
---
 .../java/com/google/gerrit/client/Dispatcher.java  |   10 +-
 .../main/java/com/google/gerrit/client/Gerrit.java |   15 ++--
 .../gerrit/client/patches/PatchBrowserPopup.java   |   30 ++++--
 .../google/gerrit/client/patches/PatchScreen.java  |  115 +++++++++++++++++---
 .../com/google/gerrit/client/ui/PatchLink.java     |   14 ++-
 5 files changed, 141 insertions(+), 43 deletions(-)

diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/Dispatcher.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/Dispatcher.java
index 7e9ffc9..d56e287 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/Dispatcher.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/Dispatcher.java
@@ -126,7 +126,7 @@ public class Dispatcher {
 
   private static void select(final String token) {
     if (token.startsWith("patch,")) {
-      patch(token, null, 0, null, null);
+      patch(token, null, 0, null, null, null);
 
     } else if (token.startsWith("change,publish,")) {
       publish(token);
@@ -285,7 +285,7 @@ public class Dispatcher {
 
   public static void patch(String token, final Patch.Key id,
       final int patchIndex, final PatchSetDetail patchSetDetail,
-      final PatchTable patchTable) {
+      final PatchTable patchTable, final PatchScreen.View popupView) {
     GWT.runAsync(new AsyncSplit(token) {
       public void onSuccess() {
         Gerrit.display(token, select());
@@ -300,7 +300,8 @@ public class Dispatcher {
               id != null ? id : Patch.Key.parse(skip(p, token)), //
               patchIndex, //
               patchSetDetail, //
-              patchTable //
+              patchTable, //
+              popupView //
           );
         }
 
@@ -310,7 +311,8 @@ public class Dispatcher {
               id != null ? id : Patch.Key.parse(skip(p, token)), //
               patchIndex, //
               patchSetDetail, //
-              patchTable //
+              patchTable, //
+              popupView //
           );
         }
 
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/Gerrit.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/Gerrit.java
index 1d27b80..27940e9 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/Gerrit.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/Gerrit.java
@@ -497,10 +497,10 @@ public class Gerrit implements EntryPoint {
     menuLeft.addInvisible(diffBar, C.menuDiff());
     addDiffLink(diffBar, CC.patchTableDiffSideBySide(), PatchScreen.Type.SIDE_BY_SIDE);
     addDiffLink(diffBar, CC.patchTableDiffUnified(), PatchScreen.Type.UNIFIED);
-    addDiffLink(diffBar, C.menuDiffCommit(), PatchScreen.TopView.COMMIT);
-    addDiffLink(diffBar, C.menuDiffPreferences(), PatchScreen.TopView.PREFERENCES);
-    addDiffLink(diffBar, C.menuDiffPatchSets(), PatchScreen.TopView.PATCH_SETS);
-    addDiffLink(diffBar, C.menuDiffFiles(), PatchScreen.TopView.FILES);
+    addDiffLink(diffBar, C.menuDiffCommit(), PatchScreen.View.COMMIT);
+    addDiffLink(diffBar, C.menuDiffPreferences(), PatchScreen.View.PREFERENCES);
+    addDiffLink(diffBar, C.menuDiffPatchSets(), PatchScreen.View.PATCH_SETS);
+    addDiffLink(diffBar, C.menuDiffFiles(), PatchScreen.View.FILES);
 
     if (signedIn) {
       m = new LinkMenuBar();
@@ -598,12 +598,12 @@ public class Gerrit implements EntryPoint {
   }
 
   private static void addDiffLink(final LinkMenuBar m, final String text,
-      final PatchScreen.TopView tv) {
+      final PatchScreen.View tv) {
     m.addItem(new LinkMenuItem(text, "") {
         @Override
         public void go() {
           if (patchScreen != null) {
-            patchScreen.setTopView(tv);
+            patchScreen.setView(tv);
           }
           AnchorElement.as(getElement()).blur();
         }
@@ -617,10 +617,11 @@ public class Gerrit implements EntryPoint {
         public void go() {
           if (patchScreen != null) {
             if (type == patchScreen.getPatchScreenType()) {
-              patchScreen.setTopView(PatchScreen.TopView.MAIN);
+              patchScreen.setView(PatchScreen.View.MAIN);
               AnchorElement.as(getElement()).blur();
             } else {
               new PatchLink("", type, patchScreen).go();
+              patchScreen.setPopupView(null);
             }
           }
         }
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchBrowserPopup.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchBrowserPopup.java
index e3fb1e9..6b18e68 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchBrowserPopup.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchBrowserPopup.java
@@ -26,38 +26,46 @@ import com.google.gwt.user.client.Window;
 import com.google.gwt.user.client.ui.FlowPanel;
 import com.google.gwt.user.client.ui.ScrollPanel;
 import com.google.gwt.user.client.ui.PopupPanel.PositionCallback;
+import com.google.gwt.user.client.ui.Widget;
 import com.google.gwtexpui.globalkey.client.GlobalKey;
 import com.google.gwtexpui.globalkey.client.HidePopupPanelCommand;
 import com.google.gwtexpui.user.client.PluginSafeDialogBox;
 
 class PatchBrowserPopup extends PluginSafeDialogBox implements
     PositionCallback, ResizeHandler {
-  private final Patch.Key callerKey;
-  private final PatchTable fileList;
+  private Patch.Key callerKey = null;
+  private PatchTable fileList = null;
+  private Widget widget;
   private final ScrollPanel sp;
   private HandlerRegistration regWindowResize;
 
-  PatchBrowserPopup(final Patch.Key pk, final PatchTable fl) {
-    super(true/* autohide */, false/* modal */);
-
+  PatchBrowserPopup(final Patch.Key pk, PatchTable fl) {
+    this(Util.M.patchSetHeader(pk.getParentKey().get()), fl);
     callerKey = pk;
     fileList = fl;
-    sp = new ScrollPanel(fileList);
+  }
+
+  PatchBrowserPopup(String header, Widget w) {
+    super(false/* autohide */, false/* modal */);
+
+    widget = w;
+    sp = new ScrollPanel(widget);
 
     final FlowPanel body = new FlowPanel();
     body.setStyleName(Gerrit.RESOURCES.css().patchBrowserPopupBody());
     body.add(sp);
 
-    setText(Util.M.patchSetHeader(callerKey.getParentKey().get()));
+    setText(header);
     setWidget(body);
     addStyleName(Gerrit.RESOURCES.css().patchBrowserPopup());
   }
 
+
   @Override
   public void setPosition(final int myWidth, int myHeight) {
     final int dLeft = (Window.getClientWidth() - myWidth) >> 1;
     final int cHeight = Window.getClientHeight();
-    final int cHeight2 = 2 * cHeight / 3;
+    final int cHeight2 = cHeight / 3;
     final int sLeft = Window.getScrollLeft();
     final int sTop = Window.getScrollTop();
 
@@ -93,7 +101,7 @@ class PatchBrowserPopup extends PluginSafeDialogBox implements
     GlobalKey.dialog(this);
     GlobalKey.addApplication(this, new HidePopupPanelCommand(0, 'f', this));
 
-    if (!fileList.isLoaded()) {
+    if (fileList != null && !fileList.isLoaded()) {
       fileList.onTableLoaded(new Command() {
         @Override
         public void execute() {
@@ -107,12 +115,12 @@ class PatchBrowserPopup extends PluginSafeDialogBox implements
   }
 
   public void open() {
-    if (!fileList.isLoaded()) {
+    if (fileList != null && !fileList.isLoaded()) {
       sp.setHeight("22px");
     }
     sp.setWidth((Window.getClientWidth() - 60) + "px");
     setPopupPositionAndShow(this);
-    if (fileList.isLoaded()) {
+    if (fileList != null && fileList.isLoaded()) {
       fileList.setRegisterKeys(true);
       fileList.movePointerTo(callerKey);
     }
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
index dcfb34a..57316c2 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
@@ -17,6 +17,7 @@ package com.google.gerrit.client.patches;
 import com.google.gerrit.client.Dispatcher;
 import com.google.gerrit.client.ErrorDialog;
 import com.google.gerrit.client.Gerrit;
+import com.google.gerrit.client.GerritConstants;
 import com.google.gerrit.client.RpcStatus;
 import com.google.gerrit.client.changes.CommitMessageBlock;
 import com.google.gerrit.client.changes.PatchTable;
@@ -32,6 +33,7 @@ import com.google.gerrit.prettify.common.PrettyFactory;
 import com.google.gerrit.reviewdb.AccountDiffPreference;
 import com.google.gerrit.reviewdb.Patch;
 import com.google.gerrit.reviewdb.PatchSet;
+import com.google.gwt.core.client.GWT;
 import com.google.gwt.core.client.Scheduler;
 import com.google.gwt.core.client.Scheduler.ScheduledCommand;
 import com.google.gwt.event.dom.client.KeyPressEvent;
@@ -41,6 +43,7 @@ import com.google.gwt.event.shared.HandlerRegistration;
 import com.google.gwt.user.client.rpc.AsyncCallback;
 import com.google.gwt.user.client.ui.FlowPanel;
 import com.google.gwt.user.client.ui.Label;
+import com.google.gwt.user.client.ui.Widget;
 import com.google.gwtexpui.globalkey.client.GlobalKey;
 import com.google.gwtexpui.globalkey.client.KeyCommand;
 import com.google.gwtexpui.globalkey.client.KeyCommandSet;
@@ -52,8 +55,9 @@ public abstract class PatchScreen extends Screen implements
 
   public static class SideBySide extends PatchScreen {
     public SideBySide(final Patch.Key id, final int patchIndex,
-        final PatchSetDetail patchSetDetail, final PatchTable patchTable) {
-      super(id, patchIndex, patchSetDetail, patchTable);
+        final PatchSetDetail patchSetDetail, final PatchTable patchTable,
+        final View popupView) {
+      super(id, patchIndex, patchSetDetail, patchTable, popupView);
     }
 
     @Override
@@ -69,8 +73,9 @@ public abstract class PatchScreen extends Screen implements
 
   public static class Unified extends PatchScreen {
     public Unified(final Patch.Key id, final int patchIndex,
-        final PatchSetDetail patchSetDetail, final PatchTable patchTable) {
-      super(id, patchIndex, patchSetDetail, patchTable);
+        final PatchSetDetail patchSetDetail, final PatchTable patchTable,
+        final View popupView) {
+      super(id, patchIndex, patchSetDetail, patchTable, popupView);
     }
 
     @Override
@@ -84,6 +89,8 @@ public abstract class PatchScreen extends Screen implements
     }
   }
 
+  public static final GerritConstants C = GWT.create(GerritConstants.class);
+
   // Which patch set id's are being diff'ed
   private static PatchSet.Id diffSideA = null;
   private static PatchSet.Id diffSideB = null;
@@ -91,8 +98,18 @@ public abstract class PatchScreen extends Screen implements
   /**
    * What should be displayed in the top of the screen
    */
-  public static enum TopView {
-    MAIN, COMMIT, PREFERENCES, PATCH_SETS, FILES
+  public static enum View {
+    MAIN (""),
+    COMMIT(C.menuDiffCommit()),
+    PREFERENCES(C.menuDiffPreferences()),
+    PATCH_SETS(C.menuDiffPatchSets()),
+    FILES(C.menuDiffFiles());
+
+    public final String title;
+
+    View(String title) {
+      this.title = title;
+    }
   }
 
   protected final Patch.Key patchKey;
@@ -110,9 +127,12 @@ public abstract class PatchScreen extends Screen implements
   private CommitMessageBlock commitMessageBlock;
   private NavLinks topNav;
   private NavLinks bottomNav;
+  private PatchBrowserPopup popup;
 
   private int rpcSequence;
   private PatchScript lastScript;
+  private View topView;
+  private View popupView;
 
   /** The index of the file we are currently looking at among the fileList */
   private int patchIndex;
@@ -131,7 +151,7 @@ public abstract class PatchScreen extends Screen implements
   }
 
   protected PatchScreen(final Patch.Key id, final int patchIndex,
-      final PatchSetDetail detail, final PatchTable patchTable) {
+      final PatchSetDetail detail, final PatchTable patchTable, View popupView) {
     patchKey = id;
     patchSetDetail = detail;
     fileList = patchTable;
@@ -164,6 +184,8 @@ public abstract class PatchScreen extends Screen implements
             setReviewedByCurrentUser(event.getValue());
           }
         });
+
+    this.popupView = popupView;
   }
 
   @Override
@@ -446,6 +468,7 @@ public abstract class PatchScreen extends Screen implements
       intralineFailure = false;
       new ErrorDialog(PatchUtil.C.intralineFailure()).show();
     }
+    setPopupView(popupView);
   }
 
   private void showPatch(final boolean showPatch) {
@@ -467,18 +490,78 @@ public abstract class PatchScreen extends Screen implements
     diffSideB = patchSetId;
   }
 
-  public void setTopView(TopView tv) {
+  public void setView(View v) {
+    if (v != topView || v == View.MAIN) {
+      setTopView(v);
+      if (v == popupView) {
+        setPopupView(null);
+      }
+    } else if (v != popupView) {
+      setTopView(View.MAIN);
+      setPopupView(v);
+    }
+  }
+
+  public void setTopView(View v) {
+    Widget w = getViewWidget(v);
     topPanel.clear();
+
+    if (w != null) {
+      topPanel.add(w);
+    }
+    topView = v;
+  }
+
+  public View getPopupView() {
+    return popupView;
+  }
+
+  public void setPopupView(View v) {
+    if (popup != null) {
+      popup.hide(false);
+    }
+
+    if (v != null) {
+      Widget w = getViewWidget(v);
+      if (w != null) {
+        if (w == fileList) {
+          popup = new PatchBrowserPopup(patchKey, fileList);
+        } else {
+          popup = new PatchBrowserPopup(v.title, w);
+        }
+        popup.open();
+      } else {
+        v = null;
+      }
+    }
+    popupView = v;
+  }
+
+  public Widget getViewWidget(View tv) {
     switch(tv) {
-      case COMMIT:      topPanel.add(commitMessageBlock);
-        break;
-      case PREFERENCES: topPanel.add(settingsPanel);
-        break;
-      case PATCH_SETS:  topPanel.add(historyTable);
-        break;
-      case FILES:       topPanel.add(fileList);
-        break;
+      case COMMIT:      return commitMessageBlock;
+      case PREFERENCES: return settingsPanel;
+      case PATCH_SETS:  return historyTable;
+      case FILES:       return fileList;
     }
+    return null;
+  }
+
+  private void popUpFileList() {
+    if (fileList == null || fileList.isAttached()) {
+      final PatchSet.Id psid = patchKey.getParentKey();
+      fileList = new PatchTable(prefs);
+      fileList.setSavePointerId("PatchTable " + psid);
+      Util.DETAIL_SVC.patchSetDetail(psid,
+          new GerritCallback<PatchSetDetail>() {
+            public void onSuccess(final PatchSetDetail result) {
+              fileList.display(result);
+            }
+          });
+    }
+
+    final PatchBrowserPopup p = new PatchBrowserPopup(patchKey, fileList);
+    p.open();
   }
 
   public class FileListCmd extends KeyCommand {
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/PatchLink.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/PatchLink.java
index 7f6de05..9d9ead7 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/PatchLink.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/PatchLink.java
@@ -25,6 +25,7 @@ public class PatchLink extends InlineHyperlink {
   protected int patchIndex;
   protected PatchSetDetail patchSetDetail;
   protected PatchTable parentPatchTable;
+  protected PatchScreen.View popupView;
 
   /**
    * @param text The text of this link
@@ -37,12 +38,13 @@ public class PatchLink extends InlineHyperlink {
   protected PatchLink(final String text, final Patch.Key patchKey,
       final int patchIndex, final String historyToken,
       final PatchSetDetail patchSetDetail,
-      final PatchTable parentPatchTable) {
+      final PatchTable parentPatchTable, final PatchScreen.View view) {
     super(text, historyToken);
     this.patchKey = patchKey;
     this.patchIndex = patchIndex;
     this.patchSetDetail = patchSetDetail;
     this.parentPatchTable = parentPatchTable;
+    this.popupView = view;
   }
 
   /**
@@ -57,7 +59,8 @@ public class PatchLink extends InlineHyperlink {
         patchScreen.getPatchIndex(), //
         Dispatcher.toPatch(type, patchScreen.getPatchKey()), //
         patchScreen.getPatchSetDetail(), //
-        patchScreen.getFileList() //
+        patchScreen.getFileList(), //
+        patchScreen.getPopupView() //
         );
   }
 
@@ -68,7 +71,8 @@ public class PatchLink extends InlineHyperlink {
         patchKey, //
         patchIndex, //
         patchSetDetail, //
-        parentPatchTable //
+        parentPatchTable, //
+        popupView //
         );
   }
 
@@ -77,7 +81,7 @@ public class PatchLink extends InlineHyperlink {
         final int patchIndex, PatchSetDetail patchSetDetail,
         PatchTable parentPatchTable) {
       super(text, patchKey, patchIndex, Dispatcher.toPatchSideBySide(patchKey),
-          patchSetDetail, parentPatchTable);
+          patchSetDetail, parentPatchTable, null);
     }
   }
 
@@ -86,7 +90,7 @@ public class PatchLink extends InlineHyperlink {
         final int patchIndex, PatchSetDetail patchSetDetail,
         PatchTable parentPatchTable) {
       super(text, patchKey, patchIndex, Dispatcher.toPatchUnified(patchKey),
-          patchSetDetail, parentPatchTable);
+          patchSetDetail, parentPatchTable, null);
     }
   }
 }
-- 
1.7.3.1

